realm/realm-library/src/main/java/io/realm/internal/ColumnIndices.java

```diff
@@ -42,147 +47,91 @@
+            new HashMap<Class<? extends RealmModel>, ColumnInfo>();
+    // Class name to ColumnInfo map. All the elements in this map should be existing in classToColumnInfoMap.
+    private final Map<String, ColumnInfo> simpleClassNameToColumnInfoMap =
+            new HashMap<String, ColumnInfo>();
 
-    /**
-     * Create a copy of the passed ColumnIndices with the specified mutability.
-     *
-     * @param other the ColumnIndices object to copy
-     * @param mutable if false the object is effectively final.
-     */
-    public ColumnIndices(ColumnIndices other, boolean mutable) {
-        this(other.schemaVersion, new HashMap<Pair<Class<? extends RealmModel>, String>, ColumnInfo>(other.classesToColumnInfo.size()), mutable);
-        for (Map.Entry<Pair<Class<? extends RealmModel>, String>, ColumnInfo> entry : other.classesToColumnInfo.entrySet()) {
-            ColumnInfo columnInfo = entry.getValue().copy(mutable);
-            Pair<Class<? extends RealmModel>, String> key = entry.getKey();
-            this.classes.put(key.first, columnInfo);
-            this.classesByName.put(key.second, columnInfo);
-            this.classesToColumnInfo.put(key, columnInfo);
-        }
-    }
+    private final RealmProxyMediator mediator;
+    // Due to the nature of Object Store's Realm::m_schema, SharedRealm's OsObjectSchemaInfo object is fixed after set.
+    private final OsSchemaInfo osSchemaInfo;
 
-    private ColumnIndices(long schemaVersion, Map<Pair<Class<? extends RealmModel>, String>, ColumnInfo> classesMap, boolean mutable) {
-        this.schemaVersion = schemaVersion;
-        this.classesToColumnInfo = classesMap;
-        this.mutable = mutable;
-        this.classes = new HashMap<>(classesMap.size());
-        this.classesByName = new HashMap<>(classesMap.size());
-    }
 
     /**
-     * Get the schema version.
+     * Create a mutable ColumnIndices initialized with the ColumnInfo objects in the passed map.
      *
-     * @return the schema version.
+     * @param mediator the {@link RealmProxyMediator} used for the corresponding Realm.
+     * @param osSchemaInfo the corresponding Realm's {@link OsSchemaInfo}.
      */
-    public long getSchemaVersion() {
-        return schemaVersion;
+    public ColumnIndices(RealmProxyMediator mediator, OsSchemaInfo osSchemaInfo) {
+        this.mediator = mediator;
+        this.osSchemaInfo = osSchemaInfo;
     }
 
     /**
-     * Returns the {@link ColumnInfo} for the passed class or ({@code null} if there is no such class).
+     * Returns the {@link ColumnInfo} for the passed class.
      *
      * @param clazz the class for which to get the ColumnInfo.
-     * @return the corresponding {@link ColumnInfo} object, or {@code null} if not found.
+     * @return the corresponding {@link ColumnInfo} object.
+     * @throws io.realm.exceptions.RealmException if the class cannot be found in the schema.
      */
-    @Nullable
+    @Nonnull
```

Comment:
We need to establish some conventions about these annotations. I see 3 options:

1) Everything should be marked with either @Nonnull or @Nullable
2) Everything is @Nonnull unless marked explicitly with @Nullable
3) Everything is @NUllable unless marked explicitly with @Nonnull

I prefer 2, but picking one and sticking to it is more important.